### ISSUE TITLE
Docker builder crashes when provisioning with "none" communicator

### ISSUE THREAD

**Issue by tylert** [hashicorp/packer#3304]
Packer crashes when using the docker builder with a communicator set to "none" and a shell provisioner.  Removing the shell provisioner fails to cause the crash.
- Packer v0.9.0
- Mac OS X 10.11.3 (Macbook Air 13" late-2013)
- Docker version 1.9.1, build a34a1d5

Crash output and templates:
https://gist.github.com/tylert/dc4d2e79d2b085e725ca

Template that causes Packer to crash:

```
{
  "builders": [
    {
      "communicator": "none",
      "export_path": "foo.tar",
      "image": "ubuntu",
      "type": "docker"
    }
  ],
  "post-processors": null,
  "provisioners": [
    {
      "inline": [
        "apt-get clean"
      ],
      "type": "shell"
    }
  ]
}
```

The following template DOES NOT cause Packer to crash:

```
{
  "builders": [
    {
      "communicator": "none",
      "export_path": "foo.tar",
      "image": "ubuntu",
      "type": "docker"
    }
  ]
}
```

Please let me know if I've missed something.

**Comment by rickard-von-essen** [comment]
Why ```post-processors: null, ```?

**Comment by tylert** [comment]
Heh.  The null post-process gets automatically added by 'packer fix'.

```
packer fix example_from_docker_builder_on_packer_website.json > temp.json
```

**Comment by tylert** [comment]
Incidentally, the null post-processor line doesn't seem to be the cause of the crash.

Using the example from the packer website docker builder page and round-tripping it through packer fix works fine:
```
{
  "builders": [
    {
      "export_path": "image.tar",
      "image": "ubuntu",
      "type": "docker"
    }
  ],
  "post-processors": null
}
```

**Comment by tylert** [comment]
Same crash with the template above occurs when using Packer 0.9.0 on Debian Jessie amd64 as well.

**Comment by cbednarski** [comment]
Thanks for opening an issue. Docker uses a default "docker" communicator so my guess is that when you set it to "none" there is no way for the shell provisioner to do anything, and packer doesn't handle this use-case.

The crash is a legitimate bug and we should certainly improve the feedback / validation for this but I think this is also a somewhat contrived example. Removing `"communicator": "none",` fixes the template. Trying to provision without a communicator is invalid; like trying to save a file to disk that is not there. For now, just don't do that. :)

**Comment by tylert** [comment]
Excellent!  As I am only *now* in the process of actually learning how to use the docker builder, the website docs have been crucial for my learning so I will likely fall into a few more of the pits that have been dug by those ahead of me.

While trying to get past a stumbling block in my practice template, I turned on "a communicator" (as the docs seemed to hint was possible) to see if it would get me past the hurdle.

(Perhaps the doc would benefit from a shell provisioner sample in the example docker template.)

Thanks again.

**Comment by tylert** [comment]
"In addition to the options listed here, a communicator can be configured for this builder."

https://github.com/mitchellh/packer/blob/174237ba2f78316ef31aa00e0d5c1cac722765a0/website/source/docs/builders/docker.html.markdown#configuration-reference

**Comment by cbednarski** [comment]
> "In addition to the options listed here, a communicator can be configured for this builder."

Yeah I don't know how you would ever use a different communicator, unless perhaps you ran an ssh server inside your container. This is possible but generally discouraged.

**Comment by borzwazie** [comment]
So, to add a little info to this bug:

I am also getting a crash, although I am not trying to use a provisioner. I have a pre-existing ISO that I want to turn into an image, and which already contains all the artifacts I need. I'm not using any provisioners at all but I still get a crash with virtualbox-iso:

panic: interface conversion: *none.comm is not packer.Communicator: missing method DownloadDir
2016/03/09 16:47:12 packer: 
2016/03/09 16:47:12 packer: goroutine 72 [running]:
2016/03/09 16:47:12 packer: panic(0x11be7c0, 0xc82027e3c0)
2016/03/09 16:47:12 packer: 	/usr/local/go/src/runtime/panic.go:464 +0x3e6
2016/03/09 16:47:12 packer: github.com/mitchellh/packer/builder/virtualbox/common.(*StepUploadVersion).Run(0xc8203134c0, 0x7f11ce314950, 0xc8202f9530, 0x0)
2016/03/09 16:47:12 packer: 	/home/username/.go/src/github.com/mitchellh/packer/builder/virtualbox/common/step_upload_version.go:18 +0xa4
2016/03/09 16:47:12 packer: github.com/mitchellh/packer/vendor/github.com/mitchellh/multistep.(*BasicRunner).Run(0xc8202e7500, 0x7f11ce314950, 0xc8202f9530)
2016/03/09 16:47:12 packer: 	/home/username/.go/src/github.com/mitchellh/packer/vendor/github.com/mitchellh/multistep/basic_runner.go:70 +0x330
2016/03/09 16:47:12 packer: github.com/mitchellh/packer/builder/virtualbox/iso.(*Builder).Run(0xc8201ca800, 0x7f11ce3143a0, 0xc8203186e0, 0x7f11ce3143e8, 0xc820313440, 0x7f11ce314418, 0xc820250080, 0x0, 0x0, 0x0, ...)
2016/03/09 16:47:12 packer: 	/home/username/.go/src/github.com/mitchellh/packer/builder/virtualbox/iso/builder.go:285 +0x19b0
2016/03/09 16:47:12 packer: github.com/mitchellh/packer/packer/rpc.(*BuilderServer).Run(0xc82024e100, 0x1, 0xc820313390, 0x0, 0x0)
2016/03/09 16:47:12 packer: 	/home/username/.go/src/github.com/mitchellh/packer/packer/rpc/builder.go:93 +0x3b9
2016/03/09 16:47:12 packer: reflect.Value.call(0x10cb760, 0x121a9d8, 0x13, 0x13d3af8, 0x4, 0xc820353ed8, 0x3, 0x3, 0x0, 0x0, ...)
2016/03/09 16:47:12 packer: 	/usr/local/go/src/reflect/value.go:435 +0x120d
2016/03/09 16:47:12 packer: reflect.Value.Call(0x10cb760, 0x121a9d8, 0x13, 0xc820353ed8, 0x3, 0x3, 0x0, 0x0, 0x0)
2016/03/09 16:47:12 packer: 	/usr/local/go/src/reflect/value.go:303 +0xb1
2016/03/09 16:47:12 packer: net/rpc.(*service).call(0xc82023a0c0, 0xc82023a080, 0xc82027c010, 0xc820248200, 0xc82024e440, 0xf39d60, 0xc82031338c, 0x18a, 0xf05260, 0xc820313390, ...)
2016/03/09 16:47:12 packer: 	/usr/local/go/src/net/rpc/server.go:383 +0x1c2
2016/03/09 16:47:12 packer: created by net/rpc.(*Server).ServeCodec
2016/03/09 16:47:12 packer: 	/usr/local/go/src/net/rpc/server.go:477 +0x49d
2016/03/09 16:47:12 ui error: Build 'virtualbox-iso' errored: unexpected EOF
2016/03/09 16:47:12 Builds completed. Waiting on interrupt barrier...
2016/03/09 16:47:12 /home/username/.go/bin/packer: plugin process exited
2016/03/09 16:47:12 machine readable: error-count []string{"1"}
2016/03/09 16:47:12 ui error: 
==> Some builds didn't complete successfully and had errors:
2016/03/09 16:47:12 machine readable: virtualbox-iso,error []string{"unexpected EOF"}
2016/03/09 16:47:12 ui error: --> virtualbox-iso: unexpected EOF
2016/03/09 16:47:12 ui: 
==> Builds finished but no artifacts were created.
2016/03/09 16:47:12 waiting for all plugin processes to complete...


Here's my json:

{
  "builders": [
    {
      "vm_name": "box1",
      "type": "virtualbox-iso",
      "http_directory": ".",
      "disk_size": 2000,
      "iso_url": "box_i686_4.0.iso",
      "iso_checksum": "4f1872fb8f1d0f095210d5269923e2b0",
      "iso_checksum_type": "md5",
      "headless": "false",
      "shutdown_command": "",
      "communicator": "none",
      "vboxmanage": [
          ["modifyvm", "{{.Name}}", "--memory", "1024"],
          ["modifyvm", "{{.Name}}", "--cpus", "2"],
          ["modifyvm", "{{.Name}}", "--nic2", "intnet"]
      ],
      "boot_command": [
        "<wait10>1<enter>y<enter><wait10><wait10><wait10><wait10>y<enter><wait10><enter>"
      ]
    }
  ]
}

**Comment by rickard-von-essen** [comment]
@borzwazie That is unrelated open a new issue. (You can work around it by setting `virtualbox_version_file` to `""`, see [virtualbox-iso - virtualbox_version_file](https://www.packer.io/docs/builders/virtualbox-iso.html#virtualbox_version_file) )

**Comment by SwampDragons** [comment]
This doesn't crash anymore; instead, the program errors out with "Error uploading script: Upload is not implemented when communicator = 'none'", which I think is clear enough to figure out what the issue is.
